File: AdvancedMapAndroid/app/src/main/java/com/carto/advancedmap/sections/basemap/VectorTileListener.java
package com.carto.advancedmap.sections.basemap;

import com.carto.datasources.LocalVectorDataSource;
import com.carto.geometry.Feature;
import com.carto.geometry.Geometry;
import com.carto.geometry.LineGeometry;
import com.carto.geometry.MultiGeometry;
import com.carto.geometry.PointGeometry;
import com.carto.geometry.PolygonGeometry;
import com.carto.graphics.Color;
import com.carto.layers.VectorLayer;
import com.carto.layers.VectorTileEventListener;
import com.carto.styles.BalloonPopupStyleBuilder;
import com.carto.styles.GeometryCollectionStyleBuilder;
import com.carto.styles.LineStyleBuilder;
import com.carto.styles.PointStyleBuilder;
import com.carto.styles.PolygonStyleBuilder;
import com.carto.ui.VectorTileClickInfo;
import com.carto.vectorelements.BalloonPopup;
import com.carto.vectorelements.GeometryCollection;
import com.carto.vectorelements.Line;
import com.carto.vectorelements.Point;
import com.carto.vectorelements.Polygon;

/**
 * Created by aareundo on 08/11/16.
 */

public class VectorTileListener extends VectorTileEventListener {

    VectorLayer layer;

    public VectorTileListener(VectorLayer layer) {
        this.layer = layer;
    }

    @Override
    public boolean onVectorTileClicked(VectorTileClickInfo clickInfo)
    {
        LocalVectorDataSource source = (LocalVectorDataSource)layer.getDataSource();

        source.clear();

        Color color = new Color((short)0, (short)100, (short)200, (short)150);

        Feature feature = clickInfo.getFeature();
        Geometry geometry = feature.getGeometry();

        PointStyleBuilder pointBuilder = new PointStyleBuilder();
        pointBuilder.setColor(color);

        LineStyleBuilder lineBuilder = new LineStyleBuilder();
        lineBuilder.setColor(color);

        PolygonStyleBuilder polygonBuilder = new PolygonStyleBuilder();
        polygonBuilder.setColor(color);

        if (geometry instanceof PointGeometry) {
            source.add(new Point((PointGeometry)geometry, pointBuilder.buildStyle()));
        } else if (geometry instanceof LineGeometry) {
            source.add(new Line((LineGeometry)geometry, lineBuilder.buildStyle()));
        }  else if (geometry instanceof PolygonGeometry) {
            source.add(new Polygon((PolygonGeometry)geometry, polygonBuilder.buildStyle()));
        } else if (geometry instanceof MultiGeometry) {

            GeometryCollectionStyleBuilder collectionBuilder = new GeometryCollectionStyleBuilder();
            collectionBuilder.setPointStyle(pointBuilder.buildStyle());
            collectionBuilder.setLineStyle(lineBuilder.buildStyle());
            collectionBuilder.setPolygonStyle(polygonBuilder.buildStyle());

            source.add(new GeometryCollection((MultiGeometry)geometry, collectionBuilder.buildStyle()));
        }

        BalloonPopupStyleBuilder builder = new BalloonPopupStyleBuilder();

        // Set a higher placement priority so it would always be visible
        builder.setPlacementPriority(10);

        String message = feature.getProperties().toString();

        BalloonPopup popup = new BalloonPopup(clickInfo.getClickPos(), builder.buildStyle(), "Click", message);

        source.add(popup);

        return true;
    }
}

File: AdvancedMapKotlin/app/src/main/java/com/carto/advanced/kotlin/sections/styles/VectorTileListener.kt
package com.carto.advanced.kotlin.sections.styles

import com.carto.datasources.LocalVectorDataSource
import com.carto.geometry.LineGeometry
import com.carto.geometry.MultiGeometry
import com.carto.geometry.PointGeometry
import com.carto.geometry.PolygonGeometry
import com.carto.graphics.Color
import com.carto.layers.VectorLayer
import com.carto.layers.VectorTileEventListener
import com.carto.styles.*
import com.carto.ui.VectorTileClickInfo
import com.carto.vectorelements.*

/**
 * Created by aareundo on 10/10/2017.
 */
class VectorTileListener(private var layer: VectorLayer) : VectorTileEventListener() {

    override fun onVectorTileClicked(clickInfo: VectorTileClickInfo): Boolean {
        val source = layer.dataSource as LocalVectorDataSource

        source.clear()

        val color = Color(0.toShort(), 100.toShort(), 200.toShort(), 150.toShort())

        val feature = clickInfo.feature
        val geometry = feature.geometry

        val pointBuilder = PointStyleBuilder()
        pointBuilder.color = color

        val lineBuilder = LineStyleBuilder()
        lineBuilder.color = color

        val polygonBuilder = PolygonStyleBuilder()
        polygonBuilder.color = color

        if (geometry is PointGeometry) {
            source.add(Point(geometry, pointBuilder.buildStyle()))
        } else if (geometry is LineGeometry) {
            source.add(Line(geometry, lineBuilder.buildStyle()))
        } else if (geometry is PolygonGeometry) {
            source.add(Polygon(geometry, polygonBuilder.buildStyle()))
        } else if (geometry is MultiGeometry) {

            val collectionBuilder = GeometryCollectionStyleBuilder()
            collectionBuilder.pointStyle = pointBuilder.buildStyle()
            collectionBuilder.lineStyle = lineBuilder.buildStyle()
            collectionBuilder.polygonStyle = polygonBuilder.buildStyle()

            source.add(GeometryCollection(geometry, collectionBuilder.buildStyle()))
        }

        val builder = BalloonPopupStyleBuilder()

        // Set a higher placement priority so it would always be visible
        builder.placementPriority = 10

        val message = feature.properties.toString()

        val popup = BalloonPopup(clickInfo.clickPos, builder.buildStyle(), "Click", message)

        source.add(popup)

        return true
    }
}